ctapipe/image/extractor.py

```diff
@@ -1285,3 +1289,400 @@ def __call__(
+            ref_pulse_shape = np.convolve(ref_pulse_shape, time_profile, "same")
+
+        integral = (
+            ref_pulse_shape.sum() * ref_sample_width_nsec / camera_sample_width_nsec
+        )
+        phase_gains, phase_shifts, phase_pz2d_shifts = [], [], []
+        for phase in range(avg_step):
+            x = np.atleast_2d(ref_pulse_shape[phase::avg_step])
+            y = deconvolve(x, 0.0, upsampling, pz)[0]
+            i_max = y.argmax()
+            start = i_max - window_shift
+            stop = start + window_width
+            if start >= 0 and stop <= y.size:
+                if leading_edge_timing:
+                    d = deconvolve(x, 0.0, upsampling, 1)[0]
+                    d_pk_idx = d.argmax()
+                    phase_pz2d_shifts.append(i_max - d_pk_idx)
+                    time = adaptive_centroid(
+                        d, d_pk_idx, leading_edge_rel_descend_limit
+                    )
+                else:
+                    time = i_max
+
+                phase_shifts.append(
+                    (time / upsampling * avg_step - ref_pulse_shape.argmax())
+                    * ref_sample_width_nsec
+                )
+                phase_gains.append(y[start:stop].sum() / integral)
+
+        if len(phase_gains) == 0:
+            raise ValueError(
+                "ref_pulse_shape is malformed - peak is not well contained"
+            )
+        gains.append(np.mean(phase_gains))
+        shifts.append(np.mean(phase_shifts))
+        pz2d_shifts.append(np.mean(phase_pz2d_shifts))
+
+    return pole_zeros, gains, shifts, pz2d_shifts
+
+
+def deconvolve(
+    waveforms: npt.ArrayLike,
+    baselines: npt.ArrayLike,
+    upsampling: int,
+    pole_zero: float,
+) -> np.ndarray:
+    """
+    Applies pole-zero deconvolution and upsampling to pixel waveforms. Use
+    `deconvolution_parameters(...)` to estimate the required `pole_zero` parameter
+    for the specific camera model.
+
+    Parameters
+    ----------
+    waveforms : ndarray
+        Waveforms stored in a numpy array.
+        Shape: (n_pix, n_samples)
+    baselines : ndarray or float
+        Baseline estimates for each pixel that are subtracted from the waveforms
+        before deconvolution.
+        Shape: (n_pix, ) or scalar
+    upsampling : int
+        Upsampling factor to use (>= 1); if > 1, the input waveforms are resampled
+        at upsampling times their original sampling rate.
+    pole_zero : float
+        Deconvolution parameter obtained from `deconvolution_parameters(...)`.
+
+    Returns
+    -------
+    deconvolved_waveforms : ndarray
+        Deconvolved and upsampled waveforms stored in a numpy array.
+        Shape: (n_pix, upsampling * n_samples)
+    """
+    deconvolved_waveforms = np.atleast_2d(waveforms) - np.atleast_2d(baselines).T
+    deconvolved_waveforms[:, 1:] -= pole_zero * deconvolved_waveforms[:, :-1]
+    deconvolved_waveforms[:, 0] = 0
+
+    def filtfilt_custom(signal, filt):
```

Comment:
This function does not seem to need a closure, so I'd define it as a private function `_filtfilt_fast` outside of this here.

I'd also add a reference to the scipy issue about the scipy version being slow on ND input.